### PR TITLE
Refactoring for reusability

### PR DESCRIPTION
**Pull request by lavalamp**
Trying to make as much of the visitor reusable as possible to support a merge operation.

It's functional but incomplete, I think?

Zero test change, 100% refactoring.

### CODE REVIEW

File: typed/traversal.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package typed
+
+import ()
```
Comment (lavalamp) [inline typed/traversal.go]: whoops, didn't mean to commit this file.

**Comment by apelisse** [comment]
LGTM

**Comment by lavalamp** [comment]
Sorry, I implemented merge here too, I should probably have sent this as a separate PR. I'll do that in a minute.

**Comment by lavalamp** [comment]
I can squash if you like before merging, but it's probably easier to review like this.

**Comment by apelisse** [comment]
I don't care about squashing, it's sort of easier to follow like this.

/lgtm

**Pull request merged**